scheduling/allocator.rb

```diff
@@ -0,0 +1,239 @@
+# frozen_string_literal: true
+
+module Scheduling::Allocator
+  def self.allocate(vm, storage_volumes, target_utilization: 0.65, distinct_storage_devices: false, allocation_state_filter: ["accepting"], host_filter: [], location_filter: [], location_preference: [])
+    request = Request.new(
+      vm.id,
+      vm.cores,
+      vm.mem_gib,
+      storage_volumes.map { _1["size_gib"] }.sum,
+      storage_volumes.size.times.zip(storage_volumes).to_h.sort_by { |k, v| v["size_gib"] * -1 },
+      distinct_storage_devices,
+      target_utilization,
+      vm.arch,
+      allocation_state_filter,
+      host_filter,
+      location_filter,
+      location_preference
+    )
+    allocation = Allocation.best_allocation(request)
+    fail "#{vm} no space left on any eligible host" unless allocation
+
+    allocation.update(vm)
+    Clog.emit("vm allocated") { {allocation: allocation.to_s, duration: Time.now - vm.created_at} }
+  end
+
+  Request = Struct.new(:vm_id, :cores, :mem_gib, :storage_gib, :storage_volumes, :distinct_storage_devices, :target_utilization,
+    :arch_filter, :allocation_state_filter, :host_filter, :location_filter, :location_preference)
+
+  class Allocation
+    attr_reader :score
+
+    def self.best_allocation(request)
+      candidate_hosts(request).map { Allocation.new(_1, request) }
+        .select { _1.is_valid }
+        .min_by { _1.score }
+    end
+
+    def self.candidate_hosts(request)
+      ds = DB[:vm_host].join(:storage_device, vm_host_id: :id)
+        .select_group(:vm_host_id, :total_cores, :used_cores, :total_hugepages_1g, :used_hugepages_1g, :location)
+        .select_append { count.function.*.as(num_storage_devices) }
+        .select_append { sum(available_storage_gib).as(available_storage_gib) }
+        .select_append { sum(total_storage_gib).as(total_storage_gib) }
+        .select_append { json_agg(json_build_object(Sequel.lit("'id'"), Sequel[:storage_device][:id], Sequel.lit("'total_storage_gib'"), total_storage_gib, Sequel.lit("'available_storage_gib'"), available_storage_gib)).order(available_storage_gib).as(storage_devices) }
+        .where(enabled: true)
+        .where(allocation_state: request.allocation_state_filter)
+        .where(arch: request.arch_filter)
+        .where { (total_hugepages_1g - used_hugepages_1g >= request.mem_gib) }
+        .where { (total_cores - used_cores >= request.cores) }
+        .having { sum(available_storage_gib) >= request.storage_gib }
+        .having { count.function.* >= (request.distinct_storage_devices ? request.storage_volumes.count : 1) }
```

Comment:
I'm asking because this is the first time I see this usage. Once we define the `count.function.*.as(num_storage_devices)`, is there a way to use its name in the `having` clause? It was kind of unintuitive to figure this out.

Reply:
I don't think that's supported in SQL

```
# select id, count(*) as c from vm_host group by id having count(*) > 1;
 id | c
----+-----
(0 rows)
```

```
# select id, count(*) as c from vm_host group by id having c > 1;
ERROR:  column "c" does not exist
LINE 1: ...ect id, count(*) as c from vm_host group by id having c > 1;
```